modules/obs-studio: optionally enable v4l2loopback

CaptainJawZ opened the pull request:
## Description of changes
created a module for obs-studio so using programs.obs-studio.enable it installs obs studio, and using the enableVirtiaulCamera setups the kernel module configurations as well as the security policy necessary for the Virtual Camera support to be enabled 
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done


<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

CaptainJawZ committed:
modules/obs-studio: optionally enable v4l2loopback

Bot-wxt1221 reviewed (changes requested):
necesary -> necessary

flexiondotorg commented:
Look like this should be include in https://github.com/NixOS/nixpkgs/pull/282903

CaptainJawZ commented:
@flexiondotorg how can I merge my code with that PR?

GaetanLepage commented:
> @flexiondotorg how can I merge my code with that PR?

My PR is stalled right now as I have not managed to write tests (that would start a VM, run OBS and check that it runs fine using OCR).

If you want to pick my changes and include them in your PR, please do not hesitate to do so.

justinas commented on nixos/modules/programs/obs-studio.nix:
```diff
@@ -0,0 +1,37 @@
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+
+    boot = mkIf cfg.enableVirtualCamera {
```
I think `kernelModules = [ "v4l2loopback" ];` is also necessary? At least from my testing.

CaptainJawZ replied:
you're right! gonna add it

CaptainJawZ committed:
add test

CaptainJawZ commented:
Alright I worked more on this, here is what I did:
- Added the kernelModule option as suggested by @justinas 
- I have added @GaetanLepage work onto my commit, joining both of us as mantainers, so now this module enables obs studio with the option to wrap plugins, enable the v4l2loopback module if virtual camera support is desired
- Finished the vm test that @GaetanLepage started, I changed it a bit, so it does several tests to ensure that v4l2loopback is properly configured, and it launches obs with a flag to start the virtual camera. As for checking up that plugins are installed, I used a plugin for capturing vulkan games, and check if its added dependencies is found on path, looking at the logs you can also see that plugins are properly loaded, but couldn't find a way to capture specifics, I hope this is good enough.

Anything else that you guys think this module needs?

GaetanLepage commented on nixos/modules/module-list.nix:
```diff
@@ -260,6 +260,8 @@
   ./programs/oddjobd.nix
   ./programs/openvpn3.nix
+  ./programs/obs-studio.nix
+  ./programs/pantheon-tweaks.nix
```
What is that `pantheon-tweaks` ?

CaptainJawZ replied:
it was a merge-conflict I fixed using the github website, let me rebase it

GaetanLepage reviewed (approved):
Great work ! Thanks for having put that effort in this module :)

GaetanLepage commented:
Also, you have to run `nixfmt-rfc-style` to format the files.

CaptainJawZ commented:
Fixed the merge conflict so it only inclues obs studio, and properly formated the nix files.

CaptainJawZ commented:
> Great work ! Thanks for having put that effort in this module :)

Thank you for providing the base! :D

GaetanLepage commented:
Now it's complaining about the missing `pantheon-tweaks.nix`. Why does that has to do with this PR ?
```
error: path '/home/runner/work/nixpkgs/nixpkgs/nixos/modules/programs/pantheon-tweaks.nix' does not exist
```

CaptainJawZ commented:
> Now it's complaining about the missing `pantheon-tweaks.nix`. Why does that has to do with this PR ?
> 
> ```
> error: path '/home/runner/work/nixpkgs/nixpkgs/nixos/modules/programs/pantheon-tweaks.nix' does not exist
> ```

My guess is that when I rebased the repo with master I must've brought that line with me, it was still mixed with my commit, when I checked files changed, I did it again from scratch and there doesnt seem to be traced of any file modification other than adding "obs-studio" to module list, so hopefully the error dissapears, sorry for the inconvenience, i'm still learning git.

GaetanLepage reviewed (changes requested):
Well done ! LGTM now.
I want to wait for another review from an experienced contributor before merging a new module.

In the meantime, please add an entry in the release changelog. I am pretty sure that this is expected for new modules.

CaptainJawZ committed:
added documentation to 24.11 changelog

CaptainJawZ commented:
Added it to the 24.11 changelog, by adding it to the nixos/doc/manual/release-notes/rl-2411.section.md but if I need to also add it to somewhere else please let me know, I didn't fully understand the contributing manual.

GaetanLepage reviewed (approved):
LGTM now !

GaetanLepage commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

- Command: `nixpkgs-review pr 312112`
- Systems: `x86_64-linux`,`aarch64-linux`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>

GaetanLepage commented:
There is a merge conflict on the release notes.

CaptainJawZ commented:
> There is a merge conflict on the release notes.

sorry, I was on a trip, I just fixed it by rebasing it to the current master

Bot-wxt1221 reviewed (approved):
LGTM

GaetanLepage reviewed (approved):
Good !

CaptainJawZ commented:
Horray! looks like it's all in order this time! thanks for the assist!
